tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_on_block.py

```diff
@@ -0,0 +1,691 @@
+    signed_block = state_transition_with_full_attestations_block(spec, fork_2_state, True, True)
+    assert fork_2_state.finalized_checkpoint.epoch == 0
+    assert fork_2_state.current_justified_checkpoint.epoch == 5
+    # Check SAFE_SLOTS_TO_UPDATE_JUSTIFIED
+    spec.on_tick(store, store.genesis_time + fork_2_state.slot * spec.config.SECONDS_PER_SLOT)
+    assert spec.compute_slots_since_epoch_start(spec.get_current_slot(store)) >= spec.SAFE_SLOTS_TO_UPDATE_JUSTIFIED
+    # Run on_block
+    yield from add_block(spec, store, signed_block, test_steps)
+    assert store.finalized_checkpoint.epoch == 0
+    assert store.justified_checkpoint.epoch == 3
+    assert store.best_justified_checkpoint.epoch == 5
+
+    # ------ fork_3_state: Create another chain to test the
+    # "Update justified if new justified is later than store justified" case
+    all_blocks = []
+    for _ in range(3):
+        next_epoch(spec, fork_3_state)
+
+    # epoch 3
+    _, signed_blocks, fork_3_state = next_epoch_with_attestations(spec, fork_3_state, True, True)
+    all_blocks += signed_blocks
+    assert fork_3_state.finalized_checkpoint.epoch == 0
+
+    # epoch 4, attest the first 5 blocks
+    _, blocks, fork_3_state = next_slots_with_attestations(spec, fork_3_state, 5, True, True)
+    all_blocks += blocks.copy()
+    assert fork_3_state.finalized_checkpoint.epoch == 0
+
+    # Propose a block at epoch 5, 5th slot
+    next_epoch(spec, fork_3_state)
+    next_slots(spec, fork_3_state, 4)
+    signed_block = state_transition_with_full_block(spec, fork_3_state, True, True)
+    all_blocks.append(signed_block.copy())
+    assert fork_3_state.finalized_checkpoint.epoch == 0
+
+    # Propose a block at epoch 6, 5th slot
+    next_epoch(spec, fork_3_state)
+    next_slots(spec, fork_3_state, 4)
+    signed_block = state_transition_with_full_block(spec, fork_3_state, True, True)
+    all_blocks.append(signed_block.copy())
+    assert fork_3_state.finalized_checkpoint.epoch == 3
+    assert fork_3_state.current_justified_checkpoint.epoch == 4
+
+    # FIXME: pending on the `on_block`, `on_attestation` fix
+    # # Apply blocks of `fork_3_state` to `store`
+    # for block in all_blocks:
+    #     if store.time < spec.compute_time_at_slot(fork_2_state, block.message.slot):
+    #         spec.on_tick(store, store.genesis_time + block.message.slot * spec.config.SECONDS_PER_SLOT)
+    #     # valid_attestations=False because the attestations are outdated (older than previous epoch)
+    #     yield from add_block(spec, store, block, test_steps, allow_invalid_attestations=False)
+
+    # assert store.finalized_checkpoint == fork_3_state.finalized_checkpoint
+    # assert (store.justified_checkpoint
+    #         == fork_3_state.current_justified_checkpoint
+    #         != store.best_justified_checkpoint)
+    # assert (store.best_justified_checkpoint
+    #         == fork_2_state.current_justified_checkpoint)
```

Comment:
Currently, the store would break since `store.justified_checkpoint is not in store.checkpoint_states` in this case.
I would have to adjust the script steps by applying blocks of `another_state` early. I wish we fix the spec issue first and then go back to clean up it.